app/src/semaphore.h

```diff
@@ -172,7 +172,30 @@ class Semaphore {
     return WaitForSingleObject(semaphore_, milliseconds) == 0;
 #else  // not windows and not mac - should be Linux.
     timespec t = internal::MsToAbsoluteTimespec(milliseconds);
-    return sem_timedwait(semaphore_, &t) == 0;
+    while (true) {
+      int result = sem_timedwait(semaphore_, &t);
+      if (result == 0) {
+        // Return success, since we successfully locked the semaphore.
+        return true;
+      }
+      switch (errno) {
+        case EINTR:
+          // Restart the wait because we were woken up spuriously.
```

Comment:
Shouldn't we reduce the amount of time that has already transpired?

Reply:
In this case, no, we should _not_ reduce the wait on the retry by the amount of time that has already transpired. The [`sem_timedwait()`](https://linux.die.net/man/3/sem_timedwait) function takes a _wall time_ as its argument, as opposed to the amount of time to wait. So you can just repeatedly specify the same `struct timespec` and if the given time is in the past then it will return immediately with `ETIMEDOUT`.

Looking at the `MsToAbsoluteTimespec()` function, which calculates the `struct timespec` specified to `sem_timedwait()`, it gets the current clock time and adds the desired timeout to it:

https://github.com/firebase/firebase-cpp-sdk/blob/c722ca77e4cbfd38f530f51c205a55d1c76bb212/app/src/time.h#L89-L95

Reply:
Got it.  Thanks!